File: content/courses/parallel-computing-introduction/codes/omp_private_firstlast.f90
program privatevar
use omp_lib

   integer ::  tid
   integer ::  nthreads=4
   integer ::  w,x,y,z

   x=10

   !$omp parallel do private(x) num_threads(nthreads)
   do i=1,nthreads
      tid=omp_get_thread_num()
      x=1000*(tid+1)
      write(*,'(a,i3,a,i6)') "Thread ",tid," gets x=",x
   enddo
   !$omp end parallel do
    
   write(*,'(a,i6)') "Outside parallel region x= ",x

   y=20;
   !$omp parallel do firstprivate(y) private(w) num_threads(nthreads)
   do i=1,nthreads
      tid=omp_get_thread_num()
      w=y*100*(tid+1)
      write(*,'(a,i3,a,i6)') "Thread ",tid," gets w=",w
   enddo
   !$omp end parallel do
    
   write(*,'(a,i6)') "Outside parallel region y= ",y

   z=30
   !$omp parallel do lastprivate(z) num_threads(nthreads)
   do i=1,nthreads
      tid=omp_get_thread_num()
      z=3000*(tid+1);
      write(*,'(a,i3,a,i6)') "Thread ",tid," gets z=",z
   enddo
   !$omp end parallel do
    
   write(*,'(a,i6)') "Outside parallel region z= ",z

end program

File: content/courses/parallel-computing-introduction/codes/omp_private_firstlast.c
#include <stdio.h>
#include <omp.h>

int main() {

    int x, y, z;
    int tid;

    int nthreads=4;

    x=10;

#pragma omp parallel for private(x) num_threads(nthreads)
    for (int i=0; i<nthreads; i++) {
        tid=omp_get_thread_num();
        x=1000*(tid+1);
        printf("Thread %d gets x= %d\n",tid, x);
    }
    
    printf("Outside parallel region x= %d\n", x);

    y=20;
#pragma omp parallel for firstprivate(y) num_threads(nthreads)
    for (int i=0; i<nthreads; i++) {
        tid=omp_get_thread_num();
        int w=y*100*(tid+1);
        printf("Thread %d gets w= %d\n",tid, w);
    }
    
    printf("Outside parallel region y= %d\n", y);

    z=30;
#pragma omp parallel for lastprivate(z) num_threads(nthreads)
    for (int i=0; i<nthreads; i++) {
        tid=omp_get_thread_num();
        z=3000*(tid+1);
        printf("Thread %d gets z= %d\n",tid, z);
    }
    
    printf("Outside parallel region z= %d\n", z);
}
